### ISSUE TITLE
Improvements in the Tomato Helpline Routing Features

### ISSUE THREAD

**Issue by radius314** [bmlt-enabled/yap#329]
The goal of this would be to check first if a root server is local to a tomato request and then use that primarily for helpline routing.  Basically only use tomato if need be.

**Comment by radius314** [comment]
@spiritualtheory I have made some changes in master.  Can you take a look at some point and let me know what you think?

**Comment by radius314** [comment]
@pjaudiomv The only difference now (I think) between `tomato_helpline_routing` and `tomato_helpline_fallback` is that the latter does an extra check.  Does it make sense to just merge this into `tomato_helpline_routing` and make one setting that does this?

**Comment by spiritualtheory** [comment]
Will look closer later - scanned through quickly and did have 2 questions : 

In the bmlt server owned - that only checks against the helpline_bmlt_setting which is an optional setting - how will it check if that setting isn’t set and just the bmlt rootserver setting is set?

Also under get meetings, the get bmltserver makes no tomato check at all so would just return the bmlt rootserver?

As for having the different settings - that was just my way of trying to not interfere with current operations and allow the original tomato helpline setting be an override setting rather than a fallback setting. The first commit was set up to really stay out of the way of existing code and operations.

So really if you don’t want an override setting, one tomato setting would work - and it just needs to check on the helpline if it’s the owned rootserver and not tomato handling the routing, otherwise meeting and helpline refer to tomato if tomato is enabled 

I’ll look later though and test functionality. The 3.40 version I reworked is working fine at the moment - just demoed it for Russia fd.

**Comment by radius314** [comment]
> Also under get meetings, the get bmltserver makes no tomato check at all so would just return the bmlt rootserver?

If you want to use tomato you would set that as your root server.

**Comment by radius314** [comment]
> As for having the different settings - that was just my way of trying to not interfere with current operations and allow the original tomato helpline setting be an override setting rather than a fallback setting. The first commit was set up to really stay out of the way of existing code and operations.

So really if you don’t want an override setting, one tomato setting would work - and it just needs to check on the helpline if it’s the owned rootserver and not tomato handling the routing, otherwise meeting and helpline refer to tomato if tomato is enabled

I don't want to create technical debt.  I understand your desire to not interfere and appreciate it, but we also want to not make more settings that are unique to one specific use-case.

**Comment by radius314** [comment]
> In the bmlt server owned - that only checks against the helpline_bmlt_setting which is an optional setting - how will it check if that setting isn’t set and just the bmlt rootserver setting is set?

I pushed a change to address this.

**Comment by spiritualtheory** [comment]
> > Also under get meetings, the get bmltserver makes no tomato check at all so would just return the bmlt rootserver?
> 
> If you want to use tomato you would set that as your root server.

But if you set tomato as your rootserver without setting a separate helpline rootserver, it breaks the check for bmlt owned

A fix for this would just be a simple if tomato set, return tomato url, else return global bmlt

**Comment by radius314** [comment]
You need to set two different ones for that use case.

**Comment by spiritualtheory** [comment]
> > As for having the different settings - that was just my way of trying to not interfere with current operations and allow the original tomato helpline setting be an override setting rather than a fallback setting. The first commit was set up to really stay out of the way of existing code and operations.
> 
> > So really if you don’t want an override setting, one tomato setting would work - and it just needs to check on the helpline if it’s the owned rootserver and not tomato handling the routing, otherwise meeting and helpline refer to tomato if tomato is enabled
> 
> I don't want to create technical debt. I understand your desire to not interfere and appreciate it, but we also want to not make more settings that are unique to one specific use-case.

Understood - though the use case really was for fixing having tomato breaking the yap routing and panel. I did expect my original method to be temporary - just was the simplest way to implement without breaking anything else

**Comment by radius314** [comment]
As far as I know what's in master now allows you accomplish exactly what you desire to occur.  If that is not the case, then please share the exact repro steps.

**Comment by spiritualtheory** [comment]
> You need to set two different ones for that use case.

Just use the same tomato routing setting you are checking against for the helpline

Really just semantics at this point - whether you put in separate rootservers in config or one rootserver and then turn tomato on or off.

**Comment by radius314** [comment]
Correct, I don't want to introduce a new setting for something that can be accomplished with an existing one.  Just waiting to hear back from @pjaudiomv if he thinks merging the feature you proposed into the `tomato_helpline_routing` feature is a viable change.  I don't think many service bodies are using this feature, so it's not too disruptive of a change and shouldn't be breaking.

**Comment by spiritualtheory** [comment]
Btw - thank you for working with me on this - I have enjoyed it immensely and look forward to learning more and hopefully contributing more productively and more compatible with existing formatting, styling, and base code.

**Comment by pjaudiomv** [comment]
One setting to rule them all, sounds good

**Comment by spiritualtheory** [comment]
So far had to correct a few things - the admin panel wasn't authorizing, the tomato meeting search wasn't catching as referenced above, and lat/long was reversed on one of the sprintf's  - also, the BMLT owned was not correctly verifying. Fixed those. However, running into the YAP volunteer routing isn't working - it just tries redirecting to the helpline which makes for an infinite loop since that's the helpline being called in the first place. Continuing to plug away at it.

**Comment by radius314** [comment]
Can you provide more specific information?

**Comment by spiritualtheory** [comment]
Admin Panel - switched all the auth functions to point to getAdminBMLTRootServer
Tomato Meeting Search - Added if/else to getBMLTRootServer to check for tomato_helpline_routing
isBMLTServerOwned - changed the manner the return compares, also fixed the mismatched lat/long ordering (showed long/lat, then referenced lat/long)

**Comment by spiritualtheory** [comment]
Everything works as far as admin panel, tomato meeting search on/off, and finding the correct service body to route to - however, thinking the tomato_helpline_routing setting is used elsewhere outside of functions and that is what is causing the issue here.

**Comment by radius314** [comment]
I was asking about the issues that were occurring.  I would ask that you
send over individual pull requests for these fixes and explain each change.

**Comment by spiritualtheory** [comment]
Not allowing me to push - saying I don't have permission

**Comment by radius314** [comment]
You need no permission to push your own fork.

**Comment by spiritualtheory** [comment]
Forgive my learning curve with github and git desktop - I re-cloned the 3.5.0 beta 3 branch to catch your updates while I'm sure there was a much better way - then it would not let me push commits. So I manually uploaded the files into my fork and did a pull from there.

So current status is it works except for the YAP forwarding to volunteers just sits on hold music, makes no calls out to the volunteers (checked twilio, plus I didn't get a call as a 24/7 volunteer)

**Comment by radius314** [comment]
available for testing in yap 3.5.0-beta4

**Comment by spiritualtheory** [comment]
Tested as working
Called non-local with tomato on
Called yap local volunteer with tomato on
Tested admin panel with tomato on
Called non-local with tomato off (correctly did not route)
Called yap local volunteer with tomato off
Tested admin panel with tomato off

Meeting search does not hit tomato now, but you said you were pulling that from this to keep it separate. I added it manually just to what’s on the NNY server and it works again. Will look to simply add this on its own commit cleanly soon.

Looks like this feature is otherwise set
